internal/konnect/node_agent_test.go

```diff
@@ -437,10 +439,116 @@ func TestNodeAgent_ControllerNodeStatusGetsUpdatedOnStatusNotification(t *testin
 
 				return true
 			}, time.Second, time.Millisecond)
+
+			expectedNodesUpdatesCount++
+			require.Equal(t, expectedNodesUpdatesCount, nodeClient.NodesUpdatesCount(), "expected only one more node update")
 		})
 	}
 }
 
+func TestNodeAgent_ControllerNodeStatusGetsUpdatedOnlyWhenItChanges(t *testing.T) {
+	nodeClient := newMockNodeClient(nil)
+	configStatusQueue := newMockConfigStatusNotifier()
+	gatewayClientsChangesNotifier := newMockGatewayClientsNotifier()
+
+	nodeAgent := konnect.NewNodeAgent(
+		testHostname,
+		testKicVersion,
+		konnect.DefaultRefreshNodePeriod,
+		logr.Discard(),
+		nodeClient,
+		configStatusQueue,
+		newMockGatewayInstanceGetter(nil),
+		gatewayClientsChangesNotifier,
+		newMockManagerInstanceIDProvider(uuid.New()),
+	)
+
+	runAgent(t, nodeAgent)
+
+	// We'll use these two to toggle between when we want to trigger an update.
+	statusOne := clients.ConfigStatusOK
+	statusTwo := clients.ConfigStatusTranslationErrorHappened
+
+	nodesUpdatesCountEventuallyEquals := func(count int) {
+		require.Eventually(t, func() bool {
+			return nodeClient.NodesUpdatesCount() == count
+		}, time.Second, time.Millisecond)
+	}
+
+	// Notify the first status and wait for the node to be updated.
+	configStatusQueue.Notify(statusOne)
+	nodesUpdatesCountEventuallyEquals(1)
+
+	// Notify the same status again and ensure that the node wasn't updated.
+	configStatusQueue.Notify(statusOne)
+	nodesUpdatesCountEventuallyEquals(1)
+
+	// Notify the second status and ensure that the node was updated.
+	configStatusQueue.Notify(statusTwo)
+	nodesUpdatesCountEventuallyEquals(2)
+
+	// Notify the same status again and ensure that the node wasn't updated.
+	configStatusQueue.Notify(statusTwo)
+	nodesUpdatesCountEventuallyEquals(2)
+}
+
+func TestNodeAgent_TickerResetsOnEveryNodesUpdate(t *testing.T) {
+	nodeClient := newMockNodeClient(nil)
+	configStatusQueue := newMockConfigStatusNotifier()
+	gatewayClientsChangesNotifier := newMockGatewayClientsNotifier()
+
+	ticker := mocks.NewTicker()
+	const halfOfRefreshPeriod = konnect.DefaultRefreshNodePeriod / 2
+	nodeAgent := konnect.NewNodeAgent(
+		testHostname,
+		testKicVersion,
+		konnect.DefaultRefreshNodePeriod,
+		logr.Discard(),
+		nodeClient,
+		configStatusQueue,
+		newMockGatewayInstanceGetter(nil),
+		gatewayClientsChangesNotifier,
+		newMockManagerInstanceIDProvider(uuid.New()),
+		konnect.WithRefreshTicker(ticker),
+	)
+
+	runAgent(t, nodeAgent)
+
+	t.Run("config status notification", func(t *testing.T) {
+		// Let half of the period pass.
+		ticker.Add(halfOfRefreshPeriod)
+
+		// Trigger update with config status notification.
+		configStatusQueue.Notify(clients.ConfigStatusApplyFailed)
+		require.Eventually(t, func() bool { return nodeClient.NodesUpdatesCount() != 1 }, time.Second, time.Nanosecond)
```

Comment:
I'm not saying we change it here but I do see that this might cause contention on weaker (i.e. CI) machines and cause the test to fail because other CPU cores couldn't pick up work because we were in a busy loop here. `time.Microsecond` might be a better choice.